### DISCUSSION TITLE
How to get token

### DISCUSSION THREAD

**Discussion by pdot72** [rubrikinc/rubrik-powershell-sdk#15] [General]
When I do the connect-rsc. How do I get the token ?
With the older version connect-polaris I was able to use the $global variable to get the token.

**Comment by markcoletta** [comment]
Per the project page readme 
🔌 Connecting to RSC
After that, you can connect to RSC with:

Connect-Rsc
It will see that there is an encrypted SA file in the default location, use it to retrieve an access token from the API server, and store it in the session state. Any subsequent calls to Connect-Rsc will use the same token.

**Reply by pdot72** [reply]
Hi thanks for the reply. 
so after i connect you are saying there is a file that stores this token? Where is this file located.?
Before with connect-polaris the token was in the $global variable and i could have uaed it to say do and invoke-webrequest and pass this token.

**Reply by guirava** [reply]
`Connect-Rsc` uses that file to generate an access token. The access token is stored in the session. Not documented: you can retrieve the connection object with `$conn=(Get-Variable RscConnectionClient).Value` 
The reason we haven't documented it is because we have the goal of designing the SDK such that you would never need to manipulate access tokens yourself (being security-minded as we are). 
Would you mind sharing what is your use case for needing a clear-text access token ?

**Reply by pdot72** [reply]
Hi 
I saw that already but that is not a token to be used as a header in a invoke-restmethod . I use this to pass my gql query.
I know there is an sdk to do this but i don't use that.

**Reply by guirava** [reply]
You mention `invoke-restmethod` : the token generated by the RSC SDK is valid with the RSC API , but not with the CDM API. They are 2 different APIs with different types of access tokens. If you're using the CDM API, you want to retrieve the access token with it.

**Reply by pdot72** [reply]
So with the other sdk, polaris sdk when using the connect-polaris it  stores the token in a global variable and with that i am able to build my body and use the invoke-restmethod with the header which includes the token. I would have expected the same.

**Reply by guirava** [reply]
oh so you are not trying to use the other SDK. With this SDK, you don't have to manipulate the access token at all. Just call `Invoke-Rsc` with your query, for example:
```powershell
Invoke-Rsc -Query "query Clusters{clusterConnection(sortBy:RegisteredAt){nodes{id}}}"
```

**Reply by pdot72** [reply]
I prefer my way where i have file with gql queries and then I pass this file and the variables and i have a function which does the invoke-restmethod and then I filter out the data. this is my function. 
 ` param (
    [string] $payload,
    [Hashtable] $variables,
    [string] $pathToData
  )
     
   
  #Create hastable to add query and variables together
  $htBody = [Ordered]@{}
 
 
  #Compress Query 
  $payload = ([String]::Join(" ", ($payload.Split("`n'")))).Trim()
  #Form Body HashTable
  $htBody.Add("query", $payload)
  $htBody.Add("variables", $variables)
  #Convert Hastable to JSON
  $body = $htBody | Convertto-json -depth 100 -compress

  #Execute Request
  $response = Invoke-restmethod -body $body -headers $headers -ContentType $Type -method post -uri $uri`
if ($pathToData) {
    $filters = $pathToData.split('.')
    foreach ($filter in $filters) {
      $response = $response.$filter
    }
  }
  return $response
  

`
So is there any way of getting the token?

**Comment by guirava** [comment]
Thank you for your feedback, 
you can now retrieve the access token from the PowerShell session with ` $global:rscconnectionclient.AccessToken`

**Reply by pdot72** [reply]
Hi I tried this after doing a connect-rsc
I still see nothing from this variable 
`$global:rscconnectionclient.AccessToken`

**Reply by pdot72** [reply]
Sorry i just updated the sdk and it works all good thanks
